File: 3903-smallest-stable-index-i/3903-smallest-stable-index-i.java
class Solution {
    public int firstStableIndex(int[] nums, int k) {
        int n = nums.length;

        int[] maxLeft = new int[n];
        int[] minRight = new int[n];

        for(int i = 0; i < n; i++) {
            if(i == 0) {
                maxLeft[i] = nums[i];
            } else {
                maxLeft[i] = Math.max(maxLeft[i - 1], nums[i]);
            }
        }
        
        for(int i = n - 1; i >= 0; i--) {
            if(i == n - 1) {
                minRight[i] = nums[i];
            } else {
                minRight[i] = Math.min(minRight[i + 1], nums[i]);
            }
        }

        for(int i = 0; i < n; i++) {
            int max = maxLeft[i];
            int min = minRight[i];

            if((max - min) <= k) {
                return i;
            }
        }
        
        return -1;
    }
}

File: 3903-smallest-stable-index-i/3903-smallest-stable-index-i.go
func firstStableIndex(nums []int, k int) int {
    n := len(nums)

    prefix := make([]int, n)
    suffix := make([]int, n)

    prefix[0] = nums[0]
    for i := 1; i < n; i++ {
        prefix[i] = max(prefix[i - 1], nums[i])
    }

    suffix[n - 1] = nums[n - 1]
    for i := n - 2; i >= 0; i-- {
        suffix[i] = min(suffix[i + 1], nums[i])
    }

    for i := 0; i < n; i++ {
        instabilityScore := prefix[i] - suffix[i]
        if instabilityScore <= k {
            return i
        }
    }

    return -1
}